intelmq/bots/experts/splunk_saved_search/expert.py

```diff
@@ -0,0 +1,202 @@
+# -*- coding: utf-8 -*-
+"""Splunk saved search enrichment export bot
+
+SPDX-FileCopyrightText: 2020 Linköping University <https://liu.se/>
+SPDX-License-Identifier: AGPL-3.0-or-later
+
+Searches Splunk for fields in an event and adds search results to it.
+
+This bot is quite slow, since it needs to submit a search job to
+Splunk, get the job ID, poll for the job to complete and then retrieve
+the results. If you have a high query load, run more instances of the
+bot.
+
+Parameters:
+
+    Generic IntelMQ HTTP parameters
+
+    auth_token: string, Splunk authentication token
+
+    url: string, base URL of the Splunk REST API
+
+    retry_interval: integer, optional, default 5, number of seconds to
+                    wait between polling for search results to be
+                    available
+
+    saved_search: string, name of Splunk saved search to run
+
+    search_parameters: map string->string, optional, default {},
+                       IntelMQ event fields to Splunk saved search
+                       parameters
+
+    result_fields: map string->string, optional, default {}, Splunk
+                   search result fields to IntelMQ event fields
+
+    not_found: list of strings, default [ "warn", "send" ], what to do
+               if the search returns zero results. All specified
+               actions are performed. Any reasonable combination of:
+               warn: log a warning message
+               send: send the event on unmodified
+               drop: drop the message
+
+    duplicates: list of strings, default [ "warn", "use_first",
+                "send" ], what to do if the search returns more than
+                one result. All specified actions are performed. Any
+                reasonable combination of:
+                warn: log a warning message
+                use_first: use the first search result
+                ignore: do not modify the event
+                send: send the event on
+                drop: drop the message
+
+    overwrite: bool or null, optional, default null, whether search
+               results replace existing values in the event. If null,
+               trying to set an existing field raises
+               intelmq.exceptions.KeyExists.
+
+"""
+
+try:
+    import requests
+except ImportError:
+    requests = None
+
+import intelmq.lib.utils as utils
+from intelmq.lib.bot import Bot
+from intelmq.lib.exceptions import MissingDependencyError, ConfigurationError
+import time
+
+
+class SplunkSavedSearchBot(Bot):
+
+    is_multithreadable = False
```

Comment:
Why is it not multithreadable?

Reply:
Everything in imperative languages is not multithreadable until proven otherwise, and given that the bot object holds state between invocations of `process()` I'm not comfortable enough with Python threading to promise anything.

Reply:
Where does the bot hold a state?

Reply:
`self.session` is created in `init()` and then continually reused in `process()`.

Reply:
I looks like the question if the requests session is thread-safe is not so easy to answer
- https://stackoverflow.com/questions/18188044/is-the-session-object-from-pythons-requests-library-thread-safe
- https://github.com/psf/requests/issues/1871
- https://github.com/psf/requests/issues/2766

I have read a bit, bot definitely not everything and it's not clear to me yet if the relevant issues affects us here. Anyway, that's a topic for a separate issue and does not need to be solved here. Thanks for pointing this out, that affects a few other bots as well.